packages/codelabs/src/basics/lit-html.md

```diff
@@ -0,0 +1,1139 @@
+      ${this.todos.map(todo => html`
+        <li>${todo.text} (${todo.finished ? 'Finished' : 'Unfinished')})</li>
+      `)}
+      </ol>
+
+      ${footerTemplate}
+    `;
+  }
+}
+```
+
+On the add button, we added an event listener for the `click` event. This is done by prefixing the event name with a `@`:
+
+```js
+<button @click=${this._addTodo}>
+```
+
+The value of an event listener should be a function. In this case, we reference a method on our component, which we should now implement:
+
+```js
+_addTodo() {
+  const input = this.shadowRoot.getElementById('addTodoInput');
+  const text = input.value;
+  input.value = '';
+
+  this.todos.push([{ text, finished: false }]);
+  this.requestUpdate();
+}
+```
+
+When this event handler is called, we create a new todo item and add it to the array of todos. Because we're mutating the todos array, `LitElement` is not able to pick up that we changed something. We can let `LitElement` know about this by calling the `requestUpdate` which exists on any element that extends from `LitElement`.
+
+When you click add, you should see the new element appear on the screen.
+
+This allows us to observe the power of lit-html in action. If you inspect the DOM while adding a todo item, you will see that only the new todo item is actually flashing:
+
+TODO: add img
+
+![todo](./assets/foo.png)
+
+When something in the DOM inspector flashes, it means that the browser is doing actual work to update the DOM tree. This is very expensive work, things like styles and layout need to be recalculated up and down the tree, so you want to minimize this as much as possible. Due to the way lit-html templating works it knows precisely what changed where and it will update only that part, without requiring a lot of from the developer.
+
+<detail>
+  <summary>View final result</summary>
+
+```html
+<!DOCTYPE html>
+<html>
+  <head></head>
+
+  <body>
+    <todo-app></todo-app>
+
+    <script type="module">
+      ...code sample goes here...
+    </script>
+  </body>
+</html>
+```
+
+</detail>
+
+## Reactive property changes
+
+Right now we're triggering an update manually when we make a change. This is fine for some use cases, but it can get pretty cumbersome to do this manually all the time and you will not be able to respond updated triggered by a parent element.
+
+It's better to let `LitElement` observe data changes and trigger updates for us. We can do this by defining `todos` as a property of our element.
+
+Start by adding a static properties object to our element, and add `todos` as an array property:
+
+```js
+static get properties() {
+  return {
+    todos: { type: Array }
+  };
+}
+```
+
+Because we're mutating the todos array using `push`, `LitElement` is not able to pick up the change. `LitElement` change detection is based on immutability, it only detects a change if you create a new array with your changes added. This is a common pattern in front-end to simplify data flow and make change detection easier.
```

Comment:
Maybe it can be good to explain a bit more how the static properties getter in lit-element works. For a long time, for me, it still felt a lot like magic. But then I understood that lit-element just used the static properties getter to make getters and setters, and that _calling_ those setters, sets in motion a chain of events to update. It really reduced the 'magic' feeling for me, and made it 'click' much more.

Reply:
The idea behind the basics codelab is to know enough to be functional, but not necessarily not how everything works behind the scenes. Do you think this should still fit in there, or this is more something for an advanced codelab?

Reply:
Yeah an aside would be good